### PR TITLE
Add option to specify name for downloaded file to http_file skylark rule

### PR DESCRIPTION
**Pull request by nlopezgi**
https://github.com/bazelbuild/bazel/issues/5633

### CODE REVIEW

File: tools/build_defs/repo/http.bzl [hunk 1]
```diff
@@ -55,25 +55,29 @@ def _http_archive_impl(ctx):
...
 
+_HTTP_FILE_FORBIDDEN_NAMES = ["WORKSPACE", "BUILD", "BUILD.bazel"]
+
 _HTTP_FILE_BUILD = """
 package(default_visibility = ["//visibility:public"])
-
 filegroup(
     name = "file",
-    srcs = ["downloaded"],
+    srcs = ["%s"],
```
Comment (xingao267) [inline tools/build_defs/repo/http.bzl]: where does this %s get the replacement value?

File: tools/build_defs/repo/http.bzl [hunk 2]
```diff
@@ -242,6 +246,7 @@ http_file = repository_rule(
     implementation = _http_file_impl,
     attrs = {
         "executable": attr.bool(),
+        "downloaded_file_name": attr.string(),
```
Comment (xingao267) [inline tools/build_defs/repo/http.bzl]: can you give the default value of `downloaded` here?

File: tools/build_defs/repo/http.bzl (hunk 1 above)
Comment (xingao267) [reply tools/build_defs/repo/http.bzl]: Oh, nvm saw it.

File: tools/build_defs/repo/http.bzl [hunk 3]
```diff
@@ -55,25 +55,29 @@ def _http_archive_impl(ctx):
...
-    srcs = ["downloaded"],
+    srcs = ["%s"],
 )
 """
 
 def _http_file_impl(ctx):
     """Implementation of the http_file rule."""
+    if ctx.attr.downloaded_file_name in _HTTP_FILE_FORBIDDEN_NAMES:
+      ctx.fail("%s cannot be used as downloaded_file_name in http_file" % ctx.attr.downloaded_file_name)
+    downloaded_file_name = ctx.attr.downloaded_file_name or "downloaded"
```
Comment (aehlig) [inline tools/build_defs/repo/http.bzl]: Use `default` on the attribute instead.

File: tools/build_defs/repo/http.bzl [hunk 4]
```diff
@@ -55,25 +55,29 @@ def _http_archive_impl(ctx):
...
     ctx.download(
         ctx.attr.urls,
-        "file/downloaded",
+        "file/" + downloaded_file_name,
         ctx.attr.sha256,
         ctx.attr.executable,
     )
     ctx.file("WORKSPACE", "workspace(name = \"{name}\")".format(name = ctx.name))
-    ctx.file("file/BUILD", _HTTP_FILE_BUILD)
+    ctx.file("file/BUILD", _HTTP_FILE_BUILD % downloaded_file_name)
```
Comment (aehlig) [inline tools/build_defs/repo/http.bzl]: prefer `format` over positional interpolation for long strings.

File: tools/build_defs/repo/http.bzl [hunk 5]
```diff
@@ -55,25 +55,29 @@ def _http_archive_impl(ctx):
...
 filegroup(
     name = "file",
-    srcs = ["downloaded"],
+    srcs = ["%s"],
 )
 """
 
 def _http_file_impl(ctx):
     """Implementation of the http_file rule."""
+    if ctx.attr.downloaded_file_name in _HTTP_FILE_FORBIDDEN_NAMES:
```
Comment (aehlig) [inline tools/build_defs/repo/http.bzl]: This check is not enough. What about `file`, `./WORKSPACE`,  special names like `.`, or path traversals like `../../../../.bashrc`?

File: tools/build_defs/repo/http.bzl (hunk 4 above)
Comment (nlopezgi) [reply tools/build_defs/repo/http.bzl]: done

File: tools/build_defs/repo/http.bzl (hunk 5 above)
Comment (nlopezgi) [reply tools/build_defs/repo/http.bzl]: changed strategy to verify invalid files, let me know if this looks better.

File: tools/build_defs/repo/http.bzl (hunk 3 above)
Comment (nlopezgi) [reply tools/build_defs/repo/http.bzl]: done

**Comment by edbaunton** [comment]
Is it permissible to specify paths in the `downloaded_file_name`, for example, `path/to/my/file`. If so, should the variable be named `downloaded_file_path`? Or should the block on paths be express in the docs?

**Comment by nlopezgi** [comment]
It is permissible (with current implementation) to pass a full path, will change name to downloaded_file_path, thanks for the advice.

**Comment by nlopezgi** [comment]
@aehlig ptal

**Comment by edbaunton** [comment]
Might be nice to add some tests - potentially somewhere around here https://github.com/bazelbuild/bazel/blob/1762772f482e6bdeb937b830dca9ba71f8f5edf7/src/test/shell/bazel/external_path_test.sh#L117 - to prevent regressions.

**Pull request closed without merging**